algebra/src/groups/group_vec.rs

```diff
@@ -208,15 +212,52 @@ impl<'a, G: Group> Mul<&'a G::ScalarField> for GroupVec<G> {
     }
 }
 
+// The trait PartialEq cannot be simply derived, because this would cause wrong results when
+// comparing different representations of the zero element of GroupVec<G> (which can be represented
+// both as an empty vector and as a vector of zero elements of the underlying group G).
+impl<G: Group> PartialEq<Self> for GroupVec<G> {
+    fn eq(&self, other: &Self) -> bool {
+        if self.is_zero() {
+            return other.is_zero();
+        }
+        if other.is_zero() {
+            return false;
+        }
+        if self.len() != other.len() {
+            return false;
+        }
+        for (el_self, el_other) in self.0.iter().zip(other.0.iter()) {
+            if el_self != el_other {
+                return false;
+            }
+        }
+        true
+    }
+}
+
+impl<G: Group> Eq for GroupVec<G> {}
+
 impl<G: Group> Group for GroupVec<G> {
     type ScalarField = G::ScalarField;
 
     fn zero() -> Self {
         GroupVec(vec![])
     }
 
+    // We allow multiple representations of the zero element of the `GroupVec`:
+    // - as an empty vector
```

Comment:
That is fine. Group vectors are unlimited in length, and it is always possible to extend/pad by as many zero elements as wished, without changing the meaning/interpretation of the vector. However, for comparison and other reasons it might be worth thinking about normalized representations, which drop tails of zero elements. (In particular every vector which only consists of zero elements collapses to the empty vector.)

Reply:
@nicholas-mainardi let's think about giving it a consistent meaning, as also needed by the circuit

Reply:
Right now in the gadget I am employing the empty vector too as a representation, dealing with this corner case explicitly in all the operations. Indeed, if we want that the operations must always be performed over vectors of the same length, then we cannot represent the zero element as a vector of zero, because we don't know the length of the vector to be compatible with other vectors (unless we add an ad-hoc zero function in the implementation of `GroupVec` which allows the caller to specify the length of the vector, as done for the `rand`).

However, even if we choose this as a "canonical" representation of zero, the zero element is necessarily represented also by a vector with all zero elements: indeed, the subtraction of a vector v by itself will unavoidably yield a vector will all zero elements, which will act as the zero element in case it is added with another vector v' of the same length. Luckily, this non-canonical representation does not need to be specifically handled in the implementations of the operations, thus we only need to deal with the empty vector, but on the other hand having two different representations of the zero element may be not elegant.